File: 61.rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if head == None:
            return None
        l = 1
        tail = head
        while tail.next != None:
            tail = tail.next
            l += 1
        k = k % l
        if k == 0:
            return head
        k = l - k
        tail.next = head
        for _ in range(1, k):
            head = head.next
        new_head = head.next
        head.next = None
        return new_head

File: 61.rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL) {
            return NULL;
        }
        int l = 1;
        ListNode* tail = head;
        while (tail->next) {
            tail = tail->next;
            l++;
        }
        k = k % l;
        if (k == 0) {
            return head;
        }
        k = l - k;
        tail->next = head;
        while (--k) {
            head = head->next;
        }
        ListNode* new_head = head->next;
        head->next = NULL;
        return new_head;
    }
};